Sprint 07

pavelbelenkow opened the pull request:
Sprint 7 tasks are completed

pavelbelenkow committed:
Added accessibility identifier to alert
Minor changes to AlertPresenter and QuestionFactory classes

pavelbelenkow committed:
MoviesLoading protocol is placed in a single file
NetworkRouting protocol is implemented, to which NetworkClient is signed for the ability to test downloads from other sources
Added implementation of networkClient via the constructor in MoviesLoader

pavelbelenkow committed:
Implemented for unit testing: ArrayTests and MoviesLoaderTests, which check the success/failure of taking an element by index in the array and the success/failure of loading data from stub
MovieQuizUITests is implemented for UI testing of buttons, the end of the game, the appearance of an alert at the end of the quiz and the beginning of a new round
Added accessibility identifiers for number of question label, buttons, image

pavelbelenkow committed:
MVP Refactoring:
Added a presenter for removing business logic from MQVC
Moved convert method to presenter
Moved currentQuestionIndex, questionsAmount properties to presenter
Added property presenter in MQVC
Removed data duplication by DRY principle

pavelbelenkow committed:
Logic of yes and no buttons has been moved to presenter
Duplication of code for implementation of the yes and no buttons on DRY principle has been removed

pavelbelenkow committed:
Refactoring 2:
moved QuestionFactoryDelegate methods to presenter
moved currentQuestion, correctAnswers, questionFactory, statisticService properties to presenter
presenter's call is implemented via the constructor
updated and renamed the resetQuestionIndex method to restartGame
moved showAnswerResult and showNextQuestionOrResults methods to presenter and renamed in proceedWithAnswer and proceedToNextQuestionOrResults methods
implemented didAnswer method to count the correct user answers
implemented questionFactoryLoadData method for calling data reloading in case of an error in View
updated and renamed the showAnswerResult method in MQVC to highlightImageBorder
added showQuizResults method in MQVC for showing an alert with statistics at the end of the game
implemented enableButtons, show/hideLoadingIndicator methods in MQVC for unlocking buttons after answering a question and showing/hiding the data loading indicator
Removed data duplication by DRY principle

pavelbelenkow committed:
Added MovieQuizViewControllerProtocol protocol for MQVC
Changed type of viewController property and constructor in presenter

pavelbelenkow committed:
Added MovieQuizPresenterTests for unit-testing of the presenter
Implemented function of testing conversion of the question model into a viewmodel

protos17 commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,159 @@
+            
+            viewController?.showQuizResults(model: alert)
+        } else {
+            
```
Рекомендую по проекту поубирать лишние отступы как здесь

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,29 @@
+    func testGetValueInRange() throws { // тест на успешное взятие элемента по индексу
+        
+        // Given
+        let array = [1, 1, 2, 3, 5]
```
Данные для тестов рекомендую брать уникальные, потому что здесь array[safe: 0] и array[safe: 1] будут выдавать одинаковое значение

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,29 @@
+    func testGetValueOutOfRange() throws { // тест на взятие элемента по неправильному индексу
+       
+        // Given
+        let array = [1, 1, 2, 3, 5]
```
Рекомендую здесь тоже использовать уникальные значения

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+    
+    func testGameFinish() throws {
+        sleep(2)
```
Нет необходимости ждать две секунды, ни на что не влияет, кроме времени теста, можно убрать

protos17 commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+    
+    func testAlertDismiss() throws {
+        sleep(2)
```
Тоже ни на что не влияет, можно убрать

pavelbelenkow committed:
Removed extra margins on the project
Added unique values to arrays for tests
Removed sleep functions in UI-tests
Implemented white status bar
